src/openqa_client/client.py

```diff
@@ -377,3 +377,33 @@ def get_jobs(
             # there as it only considers the jobs queried.
             jobdicts = self.find_clones(jobdicts)
         return jobdicts
+
+    def get_latest_build(
+        self,
+        group_id: int,
+        all_passed: bool = True,
+        sorted_key: Callable = float,
+        default_return: str = "0",
```

Comment:
Is it generally a reasonable default to return `"0"` is no build was found? I suppose `None` would be the more natural choice in Python. An empty string would probably also be a better choice because it evaluates to false in a boolean context.

Reply:
this function usually used in triggering logic , when you need decide which would be next BUILD value , for this obviously needs to know previous one .  And `default_return` would certainly be needed when NO runs ever happened in this job group so you need to start from something . In case we will remove default return value this will limit use case for this function to additional requirement `job group must have some previous triggers` which will force each caller implemented very similar code snippet : 
```
latest_build = get_latest_build(42)
if latest_build is None:
 latest_build = <default_value>
```

and in current case I can just write 
`latest_build = get_latest_build(42)`

or in case I want to be "natural Pythonist" I can go with : 
```
latest_build = get_latest_build(42, default_return=None)
if latest_build is None:
 latest_build = <default_value>
```

sorry I it is hard to me discuss what is "natural Python choice" and what is not , but in any case I would personally prefer to have one-liner even if it is not "natural" for Python . Is there some clear statement in Python spec denying return of some default value in case nothing was found ?

Reply:
No, I'm not happy with this, primarily because `"0"` is true in boolean context. Having a default return value is fine, having it be a string rather than `None` is *good* because it makes the typing clearer, but it should be the empty string `""` which evaluates False, not `"0"` which evaluates True.

Reply:
the whole point of having default return value is to **avoid handling situation when there is no build** . I would like to have ability to write `latest_build = get_latest_build(42)` and get very first build in sequence . Yes this does not make sense for non-sequential  builds but the whole implementation of this function in current state does not make sense for non-sequential builds so let's not dive into this again

Reply:
Keep in mind that situation when job group does not have builds is not an exception because it will happen in just created job group or in job group that does not have builds with all passed jobs

Reply:
I don't understand what the code you're envisioning would look like. What is some code that would take "0" as a response from this method and do something useful with it?

If this method returns "", all the calling code needs to do is a bool check on the response:
```
build = client.get_latest_build(42)
if build:
    <do stuff>
```
In your design, when the calling code gets the response "0" - a build that likely does not exist - and hits <do stuff>, what do you expect to happen? Why do you think it will be OK?

Reply:
You wrote "this function usually used in triggering logic", but designing a generic library function based on (only) your own intended use of it is usually a bad idea. Somebody might well just want to *know* what the latest build in a group is in order to run a query against it, for instance. Them getting a result which evaluates True but is actually a lie is not a great outcome.

Anyhow, if we make the return value in this case evaluate False, it's trivial to implement your desired use of it: `build=client.get_latest_build(42) or "0"`. We don't even really need a `default_return` arg in this case, as implementing it is so trivial.

But the converse isn't true - if we make the default return value a truth-y string that is a lie, users that actually need to know whether there *is* a latest build or not have to do the somewhat awkward:
```
build = client.get_latest_build(42)
if build != "0":
    <do stuff>
```
it's not really *hard*, but it's weird, and I don't think it's what anyone would expect. (it's also very awkward if a build called "0" *does* exist, of course).

I really would prefer just having the method return "" if it finds no build, and ditching the `default_return` arg.

Reply:
> I don't understand what the code you're envisioning would look like. What is some code that would take "0" as a response from this method and do something useful with it?

it is not code which I invented here for this library . It is something which I am currently using in several projects written in exactly same way . Something like : 

```
def get_next_build(self, group_id: int) -> str:
        """
        Increase input digit and format to string with zeros
        """
        return "%04d" % (int(self.get_latest_build(group_id)) + 1)
```


> If this method returns "", all the calling code needs to do is a bool check on the response:

and to address this I propagated default_returns to the caller, but for some reason you keep ignoring this ...

you code written above would perfectly works with :

```
build = client.get_latest_build(42, default_return="")
if build:
    <do stuff>
```

> You wrote "this function usually used in triggering logic", but designing a generic library function based on (only) your own intended use of it is usually a bad idea.

somehow you put yourself in the role of "advocate of majority"  who knows what majority needs . And me in role of a guy who cares only about his own code and don't care what happens around. I don't think we really can speak about some majority and argue what they want and what they don't want .


> Somebody might well just want to know what the latest build in a group is in order to run a query against it, for instance.

and this function will perfectly achieve this goal . Only point where we can not agree is how to correctly handle situation when NO builds were detected . And I actually letting caller to decide this , but nevertheless we keep arguing about situation when caller deciding NOT to change this value and get whatever function returns :see_no_evil: 

I don't see the point to return empty string for sure , better to switch to None in this case . Will you would agree with  `None` ?

Reply:
No, I prefer the empty string because of typing. If we make it the empty string, the return value is always a string. If we make it None, the return value is a string...or None.

My basic point here is that a default return value of `""` makes both use cases easy and in line with typical Python use (I write and read a *lot* of Python code and cannot recall ever having seen a function which returns the string "0" to indicate "I didn't find anything"). It makes the `default_return` argument unnecessary, because `build = client.get_latest_build(42) or <desiredvalue>` is just as easy as `build = client.get_latest_build(42, default_return=<desiredvalue>)`. It makes the code simpler and reduces the amount of tests we need to implement. I really don't see that it has any drawbacks.

Reply:
done

Reply:
thanks.